client/clientservice/src/client_service.cpp

```diff
@@ -46,10 +46,17 @@ void ClientService::start(const std::string& addr, unsigned num_async_threads, u
 
   clientservice_server->Wait();
 
-  LOG_INFO(logger_, "Wait for async gRPC threads to finish");
-  for (auto& t : server_threads_) {
-    t.join();
+  LOG_INFO(logger_, "Shutting down and emptying completion queues");
+  std::for_each(cqs_.begin(), cqs_.end(), [](auto& cq) { cq->Shutdown(); });
```

Comment:
Acc. to the [documentation](https://grpc.io/docs/languages/cpp/async/#shutting-down-the-server) we should be shutting down the server before the completion queues. Don't we need to do that here?

Reply:
`Wait()` only returns when that has happened. https://grpc.github.io/grpc/cpp/classgrpc_1_1_server_interface.html